src/cargo/core/profiles.rs

```diff
@@ -62,196 +63,84 @@ pub struct Profiles {
+        }
+
+        Ok(profiles)
+    }
+
+    /// Returns all the profiles defined in a workspace, verifying that
+    /// the requested profile exists in the workspace.
     pub fn new(ws: &Workspace<'_>, requested_profile: InternedString) -> CargoResult<Profiles> {
         let config = ws.config();
         let incremental = match config.get_env_os("CARGO_INCREMENTAL") {
             Some(v) => Some(v == "1"),
             None => config.build_config()?.incremental,
         };
-        let mut profiles = merge_config_profiles(ws, requested_profile)?;
+        let mut profiles = merge_config_profiles(ws, Some(requested_profile))?;
         let rustc_host = ws.config().load_global_rustc(Some(ws))?.host;
 
         let mut profile_makers = Profiles {
             incremental,
-            dir_names: Self::predefined_dir_names(),
+            dir_names: predefined_dir_names(),
             by_name: HashMap::new(),
             original_profiles: profiles.clone(),
             requested_profile,
             rustc_host,
         };
 
-        Self::add_root_profiles(&mut profile_makers, &profiles);
-
-        // Merge with predefined profiles.
-        use std::collections::btree_map::Entry;
-        for (predef_name, mut predef_prof) in Self::predefined_profiles().into_iter() {
-            match profiles.entry(InternedString::new(predef_name)) {
-                Entry::Vacant(vac) => {
-                    vac.insert(predef_prof);
-                }
-                Entry::Occupied(mut oc) => {
-                    // Override predefined with the user-provided Toml.
-                    let r = oc.get_mut();
-                    predef_prof.merge(r);
-                    *r = predef_prof;
-                }
-            }
+        for (name, profile) in root_profiles(&profiles) {
+            profile_makers.by_name.insert(name, profile);
         }
 
+        merge_predefined_profiles(&mut profiles);
+
         for (name, profile) in &profiles {
-            profile_makers.add_maker(*name, profile, &profiles)?;
+            let name = *name;
+            if let Some(dir_name) = profile.dir_name {
+                profile_makers.dir_names.insert(name, dir_name.to_owned());
+            }
+            if let Some(maker) =
+                create_profile_maker(name, profile, &profile_makers.by_name, &profiles)?
+            {
+                profile_makers.by_name.insert(name, maker);
+            }
         }
         // Verify that the requested profile is defined *somewhere*.
         // This simplifies the API (no need for CargoResult), and enforces
         // assumptions about how config profiles are loaded.
-        profile_makers.get_profile_maker(requested_profile)?;
+        get_profile_maker(&profile_makers.by_name, requested_profile)?;
         Ok(profile_makers)
     }
 
-    /// Returns the hard-coded directory names for built-in profiles.
-    fn predefined_dir_names() -> HashMap<InternedString, InternedString> {
```

Comment:
This seems like a refactor in order to reuse code between `Profiles::new` and `Proflies::all`. I would suggest commit the refactor in a commit, and the other for adding the new feature. It helps both review and tracing Git history.